ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,241 @@
+            zend_argument_value_error(3, "must be one of IntlListFormatter::WIDTH_WIDE, IntlListFormatter::WIDTH_SHORT, or IntlListFormatter::WIDTH_NARROW");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_openForType(locale, type, width, &status);
+    #else
+        if (type != ULISTFMT_TYPE_AND) {
+            zend_argument_value_error(2, "contains an unsupported type. ICU 66 and below only support IntlListFormatter::TYPE_AND");
+            RETURN_THROWS();
+        }
+
+        if (width != ULISTFMT_WIDTH_WIDE) {
+            zend_argument_value_error(3, "contains an unsupported width. ICU 66 and below only support IntlListFormatter::WIDTH_WIDE");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_open(locale, &status);
+    #endif
+
+    if (U_FAILURE(status)) {
+        intl_error_set(NULL, status, "Constructor failed", 0);
+        zend_throw_exception(IntlException_ce_ptr, "Constructor failed", 0);
+        RETURN_THROWS();
+    }
+}
+
+PHP_METHOD(IntlListFormatter, format)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    HashTable *ht;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_ARRAY_HT(ht)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (!LISTFORMATTER_OBJECT(obj)) {
+        zend_throw_exception(NULL, "ListFormatter not properly constructed", 0);
+        RETURN_FALSE;
+    }
+
+    uint32_t count = zend_hash_num_elements(ht);
+    if (count == 0) {
+        RETURN_EMPTY_STRING();
+    }
+
+    const UChar **items = (const UChar **)safe_emalloc(count, sizeof(const UChar *), 0);
+    int32_t *itemLengths = (int32_t *)safe_emalloc(count, sizeof(int32_t), 0);
+    uint32_t i = 0;
+    zval *val;
+
+    ZEND_HASH_FOREACH_VAL(ht, val) {
+        zend_string *str_val;
+        
+        str_val = zval_get_string(val);
+        
+        // Convert PHP string to UTF-16
+        UChar *ustr = NULL;
+        int32_t ustr_len = 0;
+        UErrorCode status = U_ZERO_ERROR;
+        
+        intl_convert_utf8_to_utf16(&ustr, &ustr_len, ZSTR_VAL(str_val), ZSTR_LEN(str_val), &status);
+        zend_string_release(str_val);
+
+        if (U_FAILURE(status)) {
+            // We can't use goto cleanup because items and itemLengths are incompletely allocated
+            for (uint32_t j = 0; j < i; j++) {
+                efree((void *)items[j]);
+            }
+            efree(items);
+            efree(itemLengths);
+            intl_error_set(NULL, status, "Failed to convert string to UTF-16", 0);
+            RETURN_FALSE;
+        }
+        
+        items[i] = ustr;
+        itemLengths[i] = ustr_len;
+        i++;
+    } ZEND_HASH_FOREACH_END();
+
+    UErrorCode status = U_ZERO_ERROR;
+    int32_t resultLength;
+    UChar *result = NULL;
+
+    resultLength = ulistfmt_format(LISTFORMATTER_OBJECT(obj), items, itemLengths, count, NULL, 0, &status);
```

Comment:
quick question, do you think there is a benefit of using `ulistfmt_formatStringsToResult` beside the "no need to calculate the buffer first" ? any drawback ?

Reply:
Note I m not holding the PR merge at all, just asking out of curiosity/trying to see a possible future improvement :)

Reply:
I'm not sure what would be the advantage (besides not calculating the buffer). The drawback is that it's supported since ICU 64 and PHP Intl supports 57 and up. I guess we could conditionally use it for >= ICU 64, but personally I would keep the code simple :)

Reply:
I would say this ... ulistfmt_formatStringsToResult is not necessarily interesting by itself but you can get out of it however is what is called field positions via a `UFieldPositionIterator`, going through it with `ufieldpositer_next` ... you can get nice additional infos ... but again my question was more to give food for thoughs ;)